src/Simulation/Simulators/SparseSimulator/Native/SparseSimulator.h

```diff
@@ -0,0 +1,846 @@
+			logical_qubit_id shift = index_2 - index_1;
+			_queued_operations.push_back(operation(OP::SWAP, index_1, shift, index_2));
+		}
+
+		void CSWAP(std::vector<logical_qubit_id> const& controls, logical_qubit_id index_1, logical_qubit_id index_2){
+			if (index_1 > index_2){
+				std::swap(index_2, index_1);
+			}
+			// Nothing commutes nicely with a controlled swap
+			_execute_if(controls);
+			_execute_if(index_1);
+			_execute_if(index_2);	
+
+			logical_qubit_id shift = index_2 - index_1;
+			_queued_operations.push_back(operation(OP::MCSWAP, index_1, shift, controls, index_2));
+			// If either qubit is occupied, then set them both to occupied
+			if(_occupied_qubits[index_1] || _occupied_qubits[index_2]){
+				_set_qubit_to_nonzero(index_1);
+				_set_qubit_to_nonzero(index_2);
+			}
+		}
+
+		bool M(logical_qubit_id target) {
+			// Do nothing if the qubit is known to be 0
+			if (!_occupied_qubits[target]){ 
+				return false;
+			}
+			// If we get a measurement, we take it as soon as we can
+			_execute_queued_ops(target, OP::Ry);
+			// If we measure 0, then this resets the occupied qubit register
+			if (_quantum_state->M(target)){
+				_set_qubit_to_nonzero(target);
+			} else {
+				_set_qubit_to_zero(target);
+			}
+			return _occupied_qubits[target];
+
+		}
+
+		void Reset(logical_qubit_id target) {
+			if (!_occupied_qubits[target]){ return; }
+			// If we get a measurement, we take it as soon as we can
+			 _execute_queued_ops(target, OP::Ry);
+			_quantum_state->Reset(target);
+			_set_qubit_to_zero(target);
+		}
+
+		void Assert(std::vector<Gates::Basis> axes, std::vector<logical_qubit_id> const& qubits, bool result) {
+			// Assertions will not commute well with Rx or Ry
+			for (auto qubit : qubits) {
+				if (_Rx_queue[qubit] || _Ry_queue[qubit]){
+					_execute_queued_ops(qubits, OP::Ry);
+				}
+			}
+			bool isAllZ = true;
+			bool isEmpty = true;
+			// Process each assertion by H commutation
+			for (int i = 0; i < qubits.size(); i++) { 		
+				switch (axes[i]){
+					case Gates::Basis::PauliY:
+						// HY=YH, so we switch the eigenvalue
+						if (_H_queue[qubits[i]]){
+							result ^= _H_queue[qubits[i]];
+						}
+						isAllZ = false;
+						isEmpty = false;
+						break; 
+					case Gates::Basis::PauliX:
+						// HX = ZH
+						if (_H_queue[qubits[i]]){
+							axes[i] = Gates::Basis::PauliZ;
+						} else {
+							isAllZ = false;
+						}
+						isEmpty = false;
+						break;
+					case Gates::Basis::PauliZ:
+						// HZ = XH
+						if (_H_queue[qubits[i]]){
+							axes[i] = Gates::Basis::PauliX;
+							isAllZ = false;
+						}
+						isEmpty = false;
+						break;
+					default:
+						break;
+				}
+			}
+			if (isEmpty) {
+				return;
+			}
+			// Z assertions are like phase gates
+			// If it's in release mode, it will queue them
+			// as a phase/permutation gate
+			// This means if an assert fails, it will fail
+			// at some future point, not at the point of failure
+			#if NDEBUG
```

Comment:
FYI: I assume you meant to write `#if defined(NDEBUG)` or `#ifdef NDEBUG`.

Reply:
Addressed here: https://github.com/sam-jaques/qsharp-runtime/pull/3